cmd/gateway-sia.go

```diff
@@ -0,0 +1,508 @@
+	sia := &siaObjects{
+		Address: host,
+		// RootDir uses access key directly, provides partitioning for
+		// concurrent users talking to same sia daemon.
+		RootDir:  os.Getenv("MINIO_ACCESS_KEY"),
+		TempDir:  os.Getenv("SIA_TEMP_DIR"),
+		password: os.Getenv("SIA_API_PASSWORD"),
+	}
+
+	// If Address not provided on command line or ENV, default to:
+	if sia.Address == "" {
+		sia.Address = "127.0.0.1:9980"
+	}
+
+	// If local Sia temp directory not specified, default to:
+	if sia.TempDir == "" {
+		sia.TempDir = ".sia_temp"
+	}
+
+	var err error
+	sia.TempDir, err = filepath.Abs(sia.TempDir)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create the temp directory with proper permissions.
+	// Ignore error when dir already exists.
+	_ = os.Mkdir(sia.TempDir, 0700)
+
+	log.Println(colorBlue("\nSia Gateway Configuration:"))
+	log.Println(colorBlue("  Sia Daemon API Address:") + colorBold(fmt.Sprintf(" %s\n", sia.Address)))
+	log.Println(colorBlue("  Sia Temp Directory:") + colorBold(fmt.Sprintf(" %s\n", sia.TempDir)))
+	return sia, nil
+}
+
+// Shutdown saves any gateway metadata to disk
+// if necessary and reload upon next restart.
+func (s *siaObjects) Shutdown() error {
+	return nil
+}
+
+// StorageInfo is not relevant to Sia backend.
+func (s *siaObjects) StorageInfo() (si StorageInfo) {
+	return si
+}
+
+// MakeBucket creates a new container on Sia backend.
+func (s *siaObjects) MakeBucketWithLocation(bucket, location string) error {
+	return nil
+}
+
+// GetBucketInfo gets bucket metadata.
+func (s *siaObjects) GetBucketInfo(bucket string) (bi BucketInfo, err error) {
+	// Until Sia support buckets/directories, must return here that all buckets exist.
+	bi.Name = bucket
+	return bi, nil
+}
+
+// ListBuckets will detect and return existing buckets on Sia.
+func (s *siaObjects) ListBuckets() (buckets []BucketInfo, err error) {
+	sObjs, serr := s.listRenterFiles("")
+	if serr != nil {
+		return buckets, serr
+	}
+
+	m := make(set.StringSet)
+
+	prefix := s.RootDir + "/"
+	for _, sObj := range sObjs {
+		if strings.HasPrefix(sObj.SiaPath, prefix) {
+			siaObj := strings.TrimPrefix(sObj.SiaPath, prefix)
+			idx := strings.Index(siaObj, "/")
+			if idx > 0 {
+				m.Add(siaObj[0:idx])
+			}
+		}
+	}
+
+	for _, bktName := range m.ToSlice() {
+		buckets = append(buckets, BucketInfo{
+			Name:    bktName,
+			Created: timeSentinel,
+		})
+	}
+
+	return buckets, nil
+}
+
+// DeleteBucket deletes a bucket on Sia.
+func (s *siaObjects) DeleteBucket(bucket string) error {
+	return nil
+}
+
+func (s *siaObjects) ListObjects(bucket string, prefix string, marker string, delimiter string, maxKeys int) (loi ListObjectsInfo, err error) {
+	siaObjs, siaErr := s.listRenterFiles(bucket)
+	if siaErr != nil {
+		return loi, siaErr
+	}
+
+	loi.IsTruncated = false
+	loi.NextMarker = ""
+
+	root := s.RootDir + "/"
+
+	for _, sObj := range siaObjs {
+		name := strings.TrimPrefix(sObj.SiaPath, pathJoin(root, bucket, "/"))
```

Comment:
don't we need to skip objects that occur before (and including) marker?

Reply:
No such things are not honored, we just list everything right now. So if Sia has 10,000 objects all of them will be listed simultaneously. Future optimizations will be done when backend supports paginated output.